src/Storages/MergeTree/MergeTreeIndexSimpleHnsw.h

```diff
@@ -0,0 +1,84 @@
+#pragma once
+
+#include <Storages/MergeTree/MergeTreeIndices.h>
+#include <Storages/MergeTree/MergeTreeData.h>
+#include <Storages/MergeTree/KeyCondition.h>
+#include <method/hnsw.h>
+
+#include <memory>
+#include <index.h>
+#include "object.h"
+namespace DB
+{
+
+struct MergeTreeIndexGranuleSimpleHnsw final : public IMergeTreeIndexGranule
+{
+    MergeTreeIndexGranuleSimpleHnsw(const String & index_name_, const Block & index_sample_block_);
+    MergeTreeIndexGranuleSimpleHnsw(const String & index_name_, const Block & index_sample_block_, std::shared_ptr<similarity::Hnsw<float>> index_impl_);
+    ~MergeTreeIndexGranuleSimpleHnsw() override = default;
+
+    void serializeBinary(WriteBuffer & ostr) const override;
+    void deserializeBinary(ReadBuffer & istr, MergeTreeIndexVersion version) override;
+
+    bool empty() const override { return false;}
+
+    String index_name;
+    Block index_sample_block;
+    std::shared_ptr<similarity::Hnsw<float>> index_impl;
+};
+
+struct MergeTreeIndexAggregatorSimpleHnsw final : IMergeTreeIndexAggregator
+{
+    MergeTreeIndexAggregatorSimpleHnsw(const String & index_name_, const Block & index_sample_block);
+    ~MergeTreeIndexAggregatorSimpleHnsw() override = default;
+
+    bool empty() const override { return true;}
+    MergeTreeIndexGranulePtr getGranuleAndReset() override;
+    void update(const Block & block, size_t * pos, size_t limit) override;
+
+    String index_name;
+    Block index_sample_block;
+    std::shared_ptr<similarity::Hnsw<float>> index_impl;
```

Comment:
Is there any reason, why it should be shared_ptr? It shouldn't copy anywhere. Maybe it is ok to create just `similarity::Hnsw<float>`?

Reply:
default constructor problem. we need default constructor for index because of  MergeTreeIndexGranulePtr MergeTreeIndexSimpleHnsw::createIndexGranule() const
{
    return std::make_shared<MergeTreeIndexGranuleSimpleHnsw>(index.name, index.sample_block);
}

Reply:
I see, let's add comment about this problem near this field, to avoid misunderstanding for other people in future.